WIP: AJAX authorization modals

leio10 opened the pull request:
#### :tophat: What? Why?
Currently, authorization modals are rendered in every page of a component, checking in advance the authorization status of the user for each handler related to every authorizable action defined by the component. This approach is simple and very performant, but also has drawbacks.

The main drawback of this approach is that **it limits the ability to add action buttons outside the component pages**. Using cards we can render information of any resource on a participatory space home page or even in the main home page, but we can't add a "Call to action" button, since we are not rendering all its related authorization modals. Instead, we are forced to add a link to a page handled by the component engine, where we can render a button that can use an authorization modal if needed.

This PR changes this behavior, loading the authorization modal contents using AJAX requests when the user tries to perform an unauthorized action. This way, the modal displaying is a little bit slower, but it can dynamically check the authorization status of the user for that action, so it can be used anywhere.

To be more specific, the only requirement to use the `action_authorized_link_to` and `action_authorized_button_to` helpers is to have the `current_component` defined, which is currently done in the components base controller and in some cells.

#### :pushpin: Related Issues

#### :clipboard: Subtasks
- [ ] Add a "Loading" message until the modal contents are loaded
- [ ] Fix/complete tests
- [ ] Add `CHANGELOG` entry

### :camera: Screenshots (optional)

leio10 commented:
@mrcasals @josepjaume @oriolgual  I don't know if you have already found the drawback that I comment on the description (for example, [with proposals](https://meta.decidim.org/processes/welcome)), or even if it is done deliberately to avoid users to perform actions before visiting the resource page. In my case, I'm facing this issue when trying to render some cells on the home page; I don't know if that kind of scenario is on the project roadmap.

Also, I've had problems with naming. I couldn't implement the new controller in the verifications gem because its needed to be used from the core gem. On the other hand, I couldn't give the name `authorizations` to the new controller because that route is used by the verifications gem. I've finally choosed the `modals` name for the controller, but I'm not happy with it, since future AJAX modal requests shouldn't be handled by this controller. Any name suggestion would be very appreciated :smiley:

deivid-rodriguez commented on decidim-consultations/app/views/decidim/consultations/questions/_vote_button.html.erb:
```diff
@@ -10,28 +10,24 @@
-                                      data: { disable: true },
-                                      class: "card__button button expanded success disabled",
-                                      id: "unvote_button" do %>
+      <%= button_to decidim_consultations.question_question_votes_path(question),
```
Why was this changed? The interface should be the same and the only change would be that the modal is ajaxly loaded, right?

leio10 replied:
Yes, you are right that is not strictly related to this change. Is the same fix than [this one](https://github.com/decidim/decidim/pull/3513/files) but in the consultations module. Both consultations and initiatives are not components, but participatory spaces, so they don't have authorizations. I understand that these helpers are only used to show the login modal if needed. 

For these cases, I defined the `logged_button_to` and `logged_link_to` and used them in this file and in the `initiatives` module templates. 

But in this very specific case, this is not needed, because it is only rendered if the user is signed in. I guess that is the same reason why @mrcasals [has done this](https://github.com/decidim/decidim/pull/3513/files#diff-6cc6b6a8fe819a100d862ce07d40cabfR6).

deivid-rodriguez replied:
Aha, understood! :+1:

deivid-rodriguez commented on decidim-core/app/controllers/decidim/modals_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  class ModalsController < Decidim::ApplicationController
+    helper_method :status, :authorize_action_path
+
+    def authorization
```
Regarding the naming issue you mentioned, what about changing the controller name to `AuthorizationModalsController` and this action to `show`?

leio10 replied:
:+1:

deivid-rodriguez commented on decidim-core/config/locales/ca.yml:
```diff
@@ -538,32 +564,7 @@ ca:
         close_modal: Tanca el modal
         filter: Filtre
         filter_by: Filtra per
-        unfold: Desplegar
```
I don't think this key movement is intended.

leio10 replied:
Thanks! I catched it in the spanish translation but missed it here.

deivid-rodriguez commented on decidim-core/spec/helpers/decidim/action_authorization_helper_spec.rb:
```diff
@@ -5,150 +5,147 @@
+
+        it "renders with a block" do
+          rendered = helper.action_authorized_link_to("foo", "fake_path") { "Link" }
+
```
Super tiny style note, you might want to remove this blank line for consistency with the other specs (although I think visually splitting non assertions from assertions reads better, so you might as well add a blank line to the other specs).

leio10 replied:
I agree with you that it reads better with an extra space there, and I don't see a clear guideline in other spec files, so I'm adding blank lines to every other specs

mrcasals replied:
I like having non-assertions and assertions separate, but It's a rule I don't enforce even on myself so... 🤷‍♂️ There's some inconsistency throughout the code about this.

deivid-rodriguez replied:
Yeah, since `rubocop-rspec` don't support it, it's hard to enforce. Maybe I'll do a feature request :)

deivid-rodriguez commented:
Yeah, I don't recall any loading spinners in the project either... And it doesn't look like foundation reveal has anything built-in. Maybe this thread has some ideas: https://foundation.zurb.com/forum/posts/35569-reveal-modal-loading-state.

leio10 commented:
@deivid-rodriguez @mrcasals maybe we could just take an icon from https://decidim-design.herokuapp.com/public/library/pattern-library and make it spin with CSS. I guess that `loop-circular`, `sun`, `reload` or `apperture` should work. What do you think?

leio10 commented:
@deivid-rodriguez I think I've addressed all your suggestions
@mrcasals I've added a CSS spinner borrowed from http://jedwatson.github.io/react-select/ that IMHO looks very nice, let me know if there is something you would change

deivid-rodriguez commented:
@leio10 Just tried it. It looks awesome! Screenshot below (don't worry, I added `sleep 2` in the controller, it's not that slow :smile:. As a matter of fact, on my computer the spinner is almost invisible by default).

![awesome](https://user-images.githubusercontent.com/2887858/42284854-28bb7e32-7f84-11e8-88e0-1fc23d85ea16.gif)

deivid-rodriguez commented on decidim-core/app/controllers/decidim/authorization_modals_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    layout false
+
+    def show
+      render
```
I don't think you need the `render` as long as there's an unlerlying template with the right name. I actually learnt from @oriolgual that you don't even new the action at all, as long as the template exists with the right name, although I find that style very obscure...

leio10 commented:
I hope I'm finished this PR with these last commits.

I've updated the CHANGELOG file and I've recovered the `ActionAuthorization` concern to cache authorizations results, but with some differences:
* I've used a `request.env` entry to ensure that each authorization verification is done once in each request (previously, each cell was having its own cache, so the proposals page perform the same check for each proposal). @mrcasals @deivid-rodriguez  Are you OK with this use of that variable? 
* I'm not including this concern into the components base controller, but in the application controller, in order to allow every page to use it (this is the main reason to use AJAX modals instead of prerendered modals).
* I'm including the component id in the cache key, to correctly cache results for actions with the same name but from different component.

mrcasals commented:
> Using cards we can render information of any resource on a participatory space home page or even in the main home page, but we can't add a "Call to action" button, since we are not rendering all its related authorization modals. Instead, we are forced to add a link to a page handled by the component engine, where we can render a button that can use an authorization modal if needed.

@leio10 when adding cells, we weren't sure you should be able to vote a proposal from the search page, so we didn't implement it and didn't check if that worked 😅

mrcasals commented on decidim-core/app/views/decidim/modals/authorization.html.erb:
```diff
@@ -0,0 +1,32 @@
+<button class="close-button" data-close aria-label="Close modal"
```
Little i18n here!

mrcasals commented:
@leio10 code looks really good, and the authorizations sections it's something mostly you and @deivid-rodriguez have been working on, so if you two don't see any possible error on it then I'm totally 👍 on the change.

About the `request.env` I don't see a problem on this, but let's wait for @deivid-rodriguez's comments before merging it 😄

leio10 commented:
> @leio10 when adding cells, we weren't sure you should be able to vote a proposal from the search page, so we didn't implement it and didn't check if that worked :sweat_smile:

@mrcasals I'm sorry if that seemed like a complain about the code, it was just an argument to justify the need of this PR. I'm very glad with the use of cells, and having that limitation was a little "thorn in the side" (or "stone in the shoe", as we would say in spanish). :smile:

mrcasals commented:
> I'm sorry if that seemed like a complain about the code, it was just an argument to justify the need of this PR

I didn't take it as a complaint, I just wanted to explain why this isn't currently possible 😄

deivid-rodriguez commented:
@leio10 I understand and agree with all your latest improvements. I'm just not sure if this is a good usage of `request.env`, maybe using the builtin `cache.fetch` and `cache.write` is better since this is about caching and not about the request environment. But I'm no expert in this area...

deivid-rodriguez commented:
Ok, that suggestion didn't making any sense, since built-in caching is meant to persist information across requests and this is an in-request cache. Global state is not usually a good sign but it seems necessary here unless we find a different design for this thing. So :+1:!

leio10 commented:
@deivid-rodriguez Mmm, I think I can't use the `cache` variable because it must be an in-memory hash, and in production it could try to send to Redis or memcached.

Regarding the use of `request.env`, I agree with you that it shouldn't store cached data. I've seen discussions like [this](https://stackoverflow.com/questions/660599/rails-per-request-hash) but I can't find an alternative as simple as `request.env`. I've found gems like [RequestStore](https://github.com/steveklabnik/request_store), but it seems a much complicated way to do the same thing.

I can think three possible ways from here:
* Wait for @josepjaume or @oriolgual to give us their opinion
* Use an "almost useless" (maybe there is a case or two when its useful) cache as it was before this PR
* Remove all the cache and leave the code simple and the discussion for later.

@mrcasals @deivid-rodriguez thoughts?

deivid-rodriguez commented:
Actually, Rails released [this feature](https://github.com/rails/rails/pull/29180) for this exact use case, no?

leio10 commented:
> Actually, Rails released this feature for this exact use case, no?

I like that feature, but I think is also more for the `current_...` variables. :smile: 

> Brain storming a bit here. What if we pass the action authorization cache as cell context in the base card? I think we're doing the same thing with the current_user.

I like this approach, I think is the best option :+1:

leio10 commented:
@deivid-rodriguez I've been playing for a while with cells context and things get much more complicated (the `ActionAuthorization` concern has to be different for controllers and cells because cells get the cache reference from the context, it fails when there are cells rendered inside cells, etc.). :sweat_smile: 

I would leave it as is, and would evaluate in the future the convenience of having one or more [CurrentAttributes](http://edgeapi.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html) classes to store the variables stored in the `request.env` hash. @deivid-rodriguez @mrcasals What do you think?

deivid-rodriguez commented:
Sounds good to me @leio10! :+1:

mrcasals commented:
Re: `Current` objects, I remember some discussion on this and we (@decidim/lot-core) didn't like the abstraction, but I can't find the issue/PR now...

Apart from that, everything looks good to me. @deivid-rodriguez do you mind reviewing the code? I trust you more than myself here 😄

leio10 commented:
@mrcasals the code doesn't changed since the last @deivid-rodriguez revision, I just rebased it to resolve conflicts and force to report coverage. :smile:

leio10 commented:
:tada: :tada: Hooray! Thanks @mrcasals!

mrcasals commented:
@leio10 thank *you* for your contributions to the project! 😄
